### PR TITLE
adding a diffusion conditioning

### PR DESCRIPTION
**Pull request by manmay-nakhashi**

### CODE REVIEW

**Comment by manmay-nakhashi** [comment]
@lucidrains i have made a skeleton for diffusion conditioning, we can keep building on top of this.

**Comment by manmay-nakhashi** [comment]
@lucidrains can you review this changes ?

**Comment by lucidrains** [comment]
@manmay-nakhashi thank you Manmay! will take a look later today and get it merged 🙌

File: naturalspeech2_pytorch/naturalspeech2_pytorch.py [hunk 1]
```diff
@@ -1367,8 +1402,11 @@ def forward(
             x_start = alpha * audio - sigma * pred
 
         _, ce_loss = self.codec.rq(x_start, codes)
-
-        return loss + self.rvq_cross_entropy_loss_weight * ce_loss
+        # pitch and duration loss
+        duration_loss = F.l1_loss(duration, duration_pred) if exists(duration) else duration_pred
+        pitch_loss = F.l1_loss(pitch, pitch_pred) if exists(pitch) else pitch_pred
+        
+        return loss + self.rvq_cross_entropy_loss_weight * ce_loss + duration_loss + pitch_loss
```
Comment (lucidrains) [inline naturalspeech2_pytorch/naturalspeech2_pytorch.py:1409]: if the `duration` and `pitch` targets were not given above, i'm not sure it can be summed to the scalar loss

File: naturalspeech2_pytorch/naturalspeech2_pytorch.py [hunk 2]
```diff
@@ -1253,7 +1271,12 @@ def process_prompt(self, prompt = None):
                 prompt, _, _ = self.codec(prompt, curtail_from_left = True, return_encoded = True)
 
         return prompt
-
+    def expand_encodings(self, phoneme_enc, attn, pitch):
+        expanded_dur = torch.einsum("klmn, kjm -> kjn", [attn, phoneme_enc])
+        pitch_emb = self.pitch_emb(f0_to_coarse(pitch).squeeze(1)).transpose(1,2)
```
Comment (lucidrains) [inline naturalspeech2_pytorch/naturalspeech2_pytorch.py:1276]: do you want to redo this with einops rearrange?

File: naturalspeech2_pytorch/naturalspeech2_pytorch.py [hunk 3]
```diff
@@ -1043,6 +1054,13 @@ def __init__(
         scale = 1.                             # this will be set to < 1. for better convergence when training on higher resolution images
     ):
         super().__init__()
+
+        self.phoneme_enc = PhonemeEncoder(tokenizer=tokenizer)
+        self.prompt_enc = SpeechPromptEncoder(dim_codebook=128) # 128 for encodec
```
Comment (lucidrains) [inline naturalspeech2_pytorch/naturalspeech2_pytorch.py:1059]: in the following 4 lines, all the hard coded number (128, 512, 80, 256), should be configurable from the `NaturalSpeech2` init

File: naturalspeech2_pytorch/utils/utils.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+            - avg: B, 1, T_en
+    """
+    durs_cums_ends = torch.cumsum(durs, dim=1).long()
+    durs_cums_starts = torch.nn.functional.pad(durs_cums_ends[:, :-1], (1, 0))
+    values_nonzero_cums = torch.nn.functional.pad(torch.cumsum(values != 0.0, dim=2), (1, 0))
+    values_cums = torch.nn.functional.pad(torch.cumsum(values, dim=2), (1, 0))
+
+    bs, l = durs_cums_ends.size()
+    n_formants = values.size(1)
+    dcs = durs_cums_starts[:, None, :].expand(bs, n_formants, l)
```
Comment (lucidrains) [inline naturalspeech2_pytorch/utils/utils.py:17]: try using einops `repeat` ?

**Review by lucidrains — COMMENTED** [review]
good start! left a few comments! thank you for this

**Comment by lucidrains** [comment]
@manmay-nakhashi could you also paste a small script for running this end to end?

File: naturalspeech2_pytorch/naturalspeech2_pytorch.py (hunk 1 above)
Comment (manmay-nakhashi) [reply naturalspeech2_pytorch/naturalspeech2_pytorch.py:1409]: pitch is passed in forward and gt duration will be used from aligner aln_hard , updating gt duration --> aln_hard

Comment (lucidrains) [reply naturalspeech2_pytorch/naturalspeech2_pytorch.py:1409]: @manmay-nakhashi yup, but should still handle the case where it is not passed in correctly `pitch_loss = F.l1_loss(pitch, pitch_pred) if exists(pitch) else pitch_pred # pitch_pred is not a loss`

Comment (lucidrains) [reply naturalspeech2_pytorch/naturalspeech2_pytorch.py:1409]: i think its fine just to rewrite it to `pitch_loss = F.l1_loss(pitch, pitch_pred) if exists(pitch) else 0.`

File: naturalspeech2_pytorch/naturalspeech2_pytorch.py [hunk 5]
```diff
@@ -109,7 +115,16 @@ def compute_pitch(spec, sample_rate, hop_length, pitch_fmax=640.0):
     )
     f0 = pw.stonemask(spec.astype(np.double), f0, t, sample_rate)
     return f0
-
+def f0_to_coarse(f0):
```
Comment (lucidrains) [inline naturalspeech2_pytorch/naturalspeech2_pytorch.py:118]: is there any reason `f0` would ever be in numpy?

i think we can just use torch and simplify conditionals below

although i do not understand the mathematics here, the `1127`, `700` can also be named and turned into keyword arguments for the `f0_to_coarse`

File: naturalspeech2_pytorch/naturalspeech2_pytorch.py [hunk 6]
```diff
@@ -1296,9 +1331,18 @@ def forward(
...
+        text_mask = create_mask()
+        
         prompt = self.process_prompt(prompt)
-
+        prompt_enc = self.prompt_enc(prompt)
+        phoneme_enc = self.phoneme_enc(text)
+        aln_hard, aln_soft, aln_log, aln_mas = self.aligner(phoneme_enc, text_mask, mel, mel_mask)
+        duration_pred, pitch_pred = self.duration_pitch(phoneme_enc,prompt_enc)
+        pitch = average_over_durations(pitch, aln_hard)
+        cond = self.expand_encodings(phoneme_enc, aln_mas.unsqueeze(1), pitch)
```
Comment (lucidrains) [inline naturalspeech2_pytorch/naturalspeech2_pytorch.py:1344]: nit: use einops rearrange for the unsqueeze here too

**Comment by lucidrains** [comment]
almost there, i'll circle back to this tomorrow morning! thank you Manmay!

File: naturalspeech2_pytorch/naturalspeech2_pytorch.py (hunk 5 above)
Comment (manmay-nakhashi) [reply naturalspeech2_pytorch/naturalspeech2_pytorch.py:118]: if we use this with compute_pitch it'll be in numpy because pw give numpy array as output.

Comment (lucidrains) [reply naturalspeech2_pytorch/naturalspeech2_pytorch.py]: ohh yes, you are correct, that sounds good!

File: naturalspeech2_pytorch/naturalspeech2_pytorch.py [hunk 7]
```diff
@@ -37,7 +38,7 @@
 from tqdm.auto import tqdm
 import pyworld as pw
 
-# constants
```
Comment (lucidrains) [inline naturalspeech2_pytorch/naturalspeech2_pytorch.py:40]: removed a line by accident

File: naturalspeech2_pytorch/utils/utils.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    n_formants = values.size(1)
+    dcs = repeat(durs_cums_starts, 'bs l -> bs n l', n=n_formants)
+    dce = repeat(durs_cums_ends, 'bs l -> bs n l', n=n_formants)
+
+    values_sums = (torch.gather(values_cums, 2, dce) - torch.gather(values_cums, 2, dcs)).to(values.dtype)
+    values_nelems = (torch.gather(values_nonzero_cums, 2, dce) - torch.gather(values_nonzero_cums, 2, dcs)).to(values.dtype)
+
+    avg = torch.where(values_nelems == 0.0, values_nelems, values_sums / values_nelems).to(values.dtype)
+    return avg
+def create_mask(sequence_length, max_len):
```
Comment (lucidrains) [inline naturalspeech2_pytorch/utils/utils.py:26]: nit: empty line before the `create_mask` function

**Comment by lucidrains** [comment]
@manmay-nakhashi fast response! once you get a working script (like the examples in the readme), let's 🚢  it

**Comment by manmay-nakhashi** [comment]
```
import torch
from naturalspeech2_pytorch import (
    EncodecWrapper,
    Model,
    NaturalSpeech2,
    SpeechPromptEncoder
)

# use encodec as an example

codec = EncodecWrapper()

model = Model(
    dim = 128,
    depth = 6,
    dim_prompt = 512,
    cond_drop_prob = 0.25,                  # dropout prompt conditioning with this probability, for classifier free guidance
    condition_on_prompt = True
)

# natural speech diffusion model

diffusion = NaturalSpeech2(
    model = model,
    codec = codec,
    timesteps = 1000
).cuda()

# mock raw audio data

raw_audio = torch.randn(4, 327680).cuda()
prompt = torch.randn(4, 32768).cuda()          # they randomly excised a range on the audio for the prompt during training, eventually will take care of this auto-magically
text = torch.randint(0, 100, (4, 100)).cuda()
mel_len = torch.tensor([120, 60 , 80, 70]).cuda()
mel = torch.randn(4,80,120).cuda()
text_lens = torch.tensor([100, 50 , 80, 120]).cuda()
pitch = torch.randn(4, 1, 120).cuda()
loss = diffusion(text=text,text_lens=text_lens, mel=mel, mel_len=mel_len,pitch=pitch,audio=raw_audio, prompt = prompt) # pass in the prompt
loss.backward()
```

**Comment by lucidrains** [comment]
@manmay-nakhashi is the mel derived from the raw audio?

**Comment by lucidrains** [comment]
@manmay-nakhashi ok, so we could take care of this automatically within the framework

might be the next logical step!

**Comment by lucidrains** [comment]
@manmay-nakhashi where would the pitch come from?

**Comment by lucidrains** [comment]
@manmay-nakhashi got it! i'll take care of that

ideally the researcher has to pass in as little information as possible, the rest computed, but still giving the option to pass everything in manually
